src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#include <math.h>
+
+/*
+ * Algorithm: Recursive Multiplying
+ *
+ * This algorithm is a generalization of the recursive doubling algorithm,
+ * and it has three stages. In the first stage, ranks above the nearest
+ * power of k less than or equal to comm_size collapse their data to the 
+ * lower ranks. The main stage proceeds with power-of-k ranks. In the main
+ * stage, ranks exchange data within groups of size k in rounds with 
+ * increasing distance (k, k^2, ...). Lastly, those in the main stage 
+ * disperse the result back to the excluded ranks. Setting k according 
+ * to the network hierarchy (e.g., the number of NICs in a node) can 
+ * improve performance.
+ */
+
+
+int MPIR_Allreduce_intra_recursive_multiplying(const void *sendbuf,
+                                            void *recvbuf,
+                                            MPI_Aint count,
+                                            MPI_Datatype datatype,
+                                            MPI_Op op,
+                                            MPIR_Comm * comm_ptr, 
+                                            const int k,
+                                            MPIR_Errflag_t errflag)
+{
+    int mpi_errno = MPI_SUCCESS;
+    /* Ensure the op is commutative */
+    MPIR_Assert(MPIR_Op_is_commutative(op));
+
+    int comm_size, rank, virt_rank;
+    comm_size = comm_ptr->local_size;
+    rank = comm_ptr->rank;
+    virt_rank = rank;
+    
+    /* get nearest power-of-two less than or equal to comm_size */
+    int power = (int) (log(comm_size) / log(k));
+    int pofk = (int) lround(pow(k, power));
+
+    MPIR_CHKLMEM_DECL(2);
+    void *tmp_buf;
+
+    /* need to allocate temporary buffer to store incoming data */
+    MPI_Aint true_extent, true_lb, extent;
+    MPIR_Type_get_true_extent_impl(datatype, &true_lb, &true_extent);
+    MPIR_Datatype_get_extent_macro(datatype, extent);
+    MPI_Aint single_node_data_size = extent * count - (extent - true_extent);
+    
+    MPIR_CHKLMEM_MALLOC(tmp_buf, void *, (k - 1) * count * single_node_data_size, mpi_errno,
+                        "temporary buffer", MPL_MEM_BUFFER);
+    
+    /* adjust for potential negative lower bound in datatype */
+    tmp_buf = (void *) ((char *) tmp_buf - true_lb);
+
+
+    /* copy local data into recvbuf */
+    if (sendbuf != MPI_IN_PLACE) {
+        mpi_errno = MPIR_Localcopy(sendbuf, count, datatype, recvbuf, count, datatype);
+        MPIR_ERR_CHECK(mpi_errno);
+    }
+
+    /* PRE-STAGE: ranks over pofk collapse their data into the others
+     * using reverse round-robin */
+    if (pofk < comm_size) {
+        if (rank >= pofk) {
+            /* This process is outside the main algorithm so send data */
+            int pre_dst = pofk - (rank % pofk) - 1;
```

Comment:
Would `pre_dst = rank % pofk` work? I think it is simpler.

Reply:
Yes it does work and would be simpler. I chose reverse round-robin to optimize for the case where there are only a few ranks above `pofk`. In this situation, I assume these ranks will have lower latency communicating with numerically similar ranks just on the other side of `pofk` instead of all the way to rank 0,1, etc.

For example, consider a case where you run on 14 nodes on Aurora. Assuming 6ppn, this is 84 processes. Lets say you pick k=3, so pofk=81. With my solution, the final 3 ranks will communicate with the other 3 ranks on the same node, whereas `pre_dst = rank % pofk` would create internode communication.

Reply:
Sure, that works. I wrote this based on the above logic before committing to the pre-stage work